### PR TITLE
fix: improve mouse actions

### PR DESCRIPTION
**Pull request by jrandolf-google**
This PR implements better state management of the mouse. This allows us to offer the mouse `buttons` attribute. This also improves processing speed by queuing them immediately.

### CODE REVIEW

File: packages/puppeteer-core/src/common/Input.ts
Comment (OrKoN) [inline packages/puppeteer-core/src/common/Input.ts:343]: let's add a comment explaining how it works.

File: packages/puppeteer-core/src/common/Input.ts
Comment (OrKoN) [inline packages/puppeteer-core/src/common/Input.ts:349]: ditto

File: packages/puppeteer-core/src/common/Input.ts
Comment (OrKoN) [inline packages/puppeteer-core/src/common/Input.ts:359]: ditto

File: packages/puppeteer-core/src/common/Input.ts
Comment (OrKoN) [inline packages/puppeteer-core/src/common/Input.ts:642]: let's add a test for this branch: that the error is reported

File: test/src/mouse.spec.ts
Comment (OrKoN) [inline test/src/mouse.spec.ts:262]: that looks like a potentially breaking change? was it throwing before? at least let's mark it as a non-breaking feature so that it shows up in the changelog.

Comment (jrandolf) [reply test/src/mouse.spec.ts:262]: It wasn't throwing before, but it would no-op. This makes this explicit.

Comment (jrandolf) [reply test/src/mouse.spec.ts:262]: I've set this to a fix since this is technically the correct behavior. Is that fine?

Comment (OrKoN) [reply test/src/mouse.spec.ts:262]: sgtm!

**Comment by OrKoN** [comment]
Tests are still failing. I will wait for CI to be green before taking the final look.

File: packages/puppeteer-core/src/common/Input.ts
Comment (OrKoN) [inline packages/puppeteer-core/src/common/Input.ts]: Let's say we have two move points: p1 & p2. After we start Promise.all, this.#position is equal p2. Now let's assume that that dispatchMouseEvent for p2 succeeds first and call for p1 is rejected. The this.#position is then incorrectly set to p1 whereas the mouse is at p2?

File: packages/puppeteer-core/src/common/Input.ts
Comment (OrKoN) [inline packages/puppeteer-core/src/common/Input.ts]: why don't we keep it as in the previous version? when the position is updated before all sync calls have been done?

Comment (OrKoN) [reply packages/puppeteer-core/src/common/Input.ts]: We can also rollback the position if the last dispatchMouseCall didn't succeed.

Comment (jrandolf) [reply packages/puppeteer-core/src/common/Input.ts]: I've updated the code so the state management is more transparent.

File: packages/puppeteer-core/src/common/Input.ts
Comment (jrandolf) [reply packages/puppeteer-core/src/common/Input.ts]: Obsolete

File: packages/puppeteer-core/src/common/Input.ts
Comment (OrKoN) [inline packages/puppeteer-core/src/common/Input.ts]: So is my understanding correct for the following case?

Two events are dispatched in one task: mousedown, mousemove. Then events happens in the following order:

```
1) mousdown resolves
2) previous state = current state =  (mousedown + mousemove) at line 531
3) mousemove rejects
4) state = previous state = mousedown + mousemove at line 533
```

So it looks like rollback state is not correct since it has to only include the mousedown state.

Comment (jrandolf) [reply packages/puppeteer-core/src/common/Input.ts]: Fixed.

File: packages/puppeteer-core/src/common/Input.ts
Comment (OrKoN) [inline packages/puppeteer-core/src/common/Input.ts:677]: nit: it could be simplified I think so that actions don't need to be cleared:

```
const delayPromise = typeof delay === 'number' ? ...timeout : Promise.resolve();
actions.push(delayPromise.then(() => this.up(options)));
```

**Review by OrKoN — APPROVED** [review]

File: packages/puppeteer-core/src/common/Input.ts
Comment (jrandolf) [reply packages/puppeteer-core/src/common/Input.ts:677]: Ack

File: packages/puppeteer-core/src/common/Input.ts
Comment (melink14) [inline packages/puppeteer-core/src/common/Input.ts:642]: Previously, if you called `up` and the button wasn't pressed, this function would not throw an error.

Should this have been a breaking change?

For example, the testing library I use has a `resetMouse` function which calls `up` on all buttons to make sure the mouse is in it's initial state and after this change using that function throws an error.

Comment (OrKoN) [reply packages/puppeteer-core/src/common/Input.ts:642]: @jrandolf

Comment (jrandolf) [reply packages/puppeteer-core/src/common/Input.ts:642]: So are you requesting the `up` button reset to a no-op?

Comment (melink14) [reply packages/puppeteer-core/src/common/Input.ts:642]: My main goal was just to clarify the breaking behavior of the change since I'm not an expert in this code base. 2 weeks later there might not be much to do except maybe updating the release notes so that it's easier for users to find the breaking change.

Perhaps it would be useful if the documentation should be updated to clarify the constraints on calling the methods.

As to the actual design decision, I think it comes down to user expectations. If we think most users calling `up` without calling `down` are doing it by accident then throwing an error is probably better. If not, then maybe a warning would be better (or nothing).

Since users who want to unconditionally reset mouse can use `try...catch` to avoid major problems with the error, it's fine to inconvenience them if they're in the minority imo.

Comment (OrKoN) [reply packages/puppeteer-core/src/common/Input.ts:642]: @jrandolf let's improve documentation around this (by calling out the new error more explicitly in the PR description and updating the method docs)?

Comment (thepassle) [reply packages/puppeteer-core/src/common/Input.ts:642]: This indeed seems to have broken a package of ours: https://github.com/modernweb-dev/web/issues/2262

Apparently now: "it's an error to call up if the mouse button is not actually pressed"

The code in our package looks as follows: https://github.com/modernweb-dev/web/blob/e0d362fcdbedf6a72a6c78ebc2e3a3cffa922dbc/packages/test-runner-commands/src/sendMousePlugin.ts#L170

Could a maintainer of puppeteer provide some guidance on what would be the fix for this? Apparently we just need to "actually press" the mouse button? What would that look like?

Comment (OrKoN) [reply packages/puppeteer-core/src/common/Input.ts:642]:
```
await page.mouse.down({ button: 'left' });
```

should be called before up:

```
await page.mouse.up({ button: 'left' });
```

it looks like you only send up calls without the down calls. Alternatively, it seems you can use `click` to have down/up paired up correctly.

Comment (jrandolf) [reply packages/puppeteer-core/src/common/Input.ts:642]: Their `up` calls are being used to just reset the mouse state. WDYT about having a `reset` function on the mouse? This way, the user doesn't need to keep track of the mouse state to reset it.

Comment (OrKoN) [reply packages/puppeteer-core/src/common/Input.ts:642]: @jrandolf should we have some sort of reset mechanism for the mouse/keyboard? I can see the use case for tests where it is difficult to know if the tests left the keyboard/mouse in the right state.

Comment (jrandolf) [reply packages/puppeteer-core/src/common/Input.ts:642]: Okay, I think the fact we both commented simultaneously implies we should have one. Let me create a PR.

Comment (thepassle) [reply packages/puppeteer-core/src/common/Input.ts:642]: Could you ping me when this is implemented and released? Seems like a `reset` function would be great here indeed :)

Comment (jrandolf) [reply packages/puppeteer-core/src/common/Input.ts:642]: See https://github.com/puppeteer/puppeteer/pull/10340

**Pull request merged**
